Add more MapIdentityApi endpoints

halter73 opened the pull request:
This adds the following new endpoints:

- GET /confirmEmail
- POST /resendConfirmationEmail
- POST /resetPassword
- GET /account/2fa
- POST /account/2fa
- GET /account/info
- POST /account/info

Additionally, the existing /login endpoint now accepts 2fa codes and 2fa recovery codes as part of the request body. These can be queried and regenerated from /account/2fa. The /login endpoint now also gives limited failure reasons in the form of application/problem+json instead of empty 401 responses with details such as "LockedOut", "RequiresTwoFactor", "NotAllowed" (usually because lack of email confirmation), and the generic "Failed" statuses.

Fixes #47232 (lockout support)
Fixes #47231 (reset password support)
Fixes #47230 (2fa support)
Fixes #47229 (change username and password)
Fixes #49404 (removes AddIdentityBearerToken which is no longer needed)

halter73 committed:
Add lockout support

halter73 committed:
Add email support

halter73 committed:
Add multitenency

halter73 committed:
Add 2fa

halter73 committed:
Add /account management APIs and general cleanup

AlexKovynev commented:
maybe https://github.com/dotnet/aspnetcore/issues/47232 not https://github.com/dotnet/aspnetcore/issues/47432 :)

halter73 committed:
Minor cleanup

halter73 commented:
/backport to release/8.0-preview7

halter73 committed:
Use TypedResults.Empty

halter73 committed:
Use IUserStore to set username during registration

halter73 committed:
Fix CanUsePersistentTwoFactorCookies

- Use ToDictionary when it's straightforward
- Clear ProviderInstance when adding provider type to existing TokenProviderDescriptor
- Apply API review feedback

BrennanConroy commented on src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs:
```diff
@@ -42,44 +59,64 @@ public static IEndpointConventionBuilder MapIdentityApi<TUser>(this IEndpointRou
             var user = new TUser();
-            await userManager.SetUserNameAsync(user, registration.Username);
+            await userStore.SetUserNameAsync(user, registration.Username, CancellationToken.None);
+            await emailStore.SetEmailAsync(user, registration.Email, CancellationToken.None);
```
Should we get the `HttpContext` and pass `RequestAborted` to these?

halter73 replied:
This copies Identity UI. I'm not sure it makes sense to cancel a database transaction just because the request is aborted. In practice with the EF stores, these all complete immediately after changing the in-memory representation. Only UserManager.CreateAsync below actually commits anything to the database and that doesn't accept a cancellation token.

There is the [AspNetUserManager](https://github.com/dotnet/aspnetcore/blob/a56e968c19be1275db6dc462310d723615b006a7/src/Identity/Core/src/AspNetUserManager.cs) that attaches the RequestAborted token to the UserManager whether it makes sense or not. I think that's at least all or nothing for user creation. We don't want a partially created user which might be technically possible for the preview7 backport but is fixed here by [`dcd5609` (#49498)](https://github.com/dotnet/aspnetcore/pull/49498/commits/dcd560997bcdac44e96a570cc43c197ff6e8daa0) 😆.

BrennanConroy commented on src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs:
```diff
@@ -42,44 +59,64 @@ public static IEndpointConventionBuilder MapIdentityApi<TUser>(this IEndpointRou
+            var signInManager = sp.GetRequiredService<SignInManager<TUser>>();
+
+            signInManager.PrimaryAuthenticationScheme = cookieMode == true ? IdentityConstants.ApplicationScheme : IdentityConstants.BearerScheme;
+            var isPersistent = persistCookies ?? true;
```
Do we care if `persistCookies` is set when `cookieMode` is `false` or `null`?

halter73 replied:
It doesn't make a difference.

BrennanConroy commented on src/Identity/Core/src/DTO/IdentityEndpointsJsonSerializerContext.cs:
```diff
@@ -7,6 +7,13 @@ namespace Microsoft.AspNetCore.Identity.DTO;
+[JsonSerializable(typeof(InfoRequest))]
+[JsonSerializable(typeof(InfoResponse))]
+[JsonSerializable(typeof(TwoFactorRequest))]
+[JsonSerializable(typeof(TwoFactorResponse))]
```
Would any of these classes benefit from being structs? I think minimal tries to avoid boxing when it can, and Json avoids boxing sometimes too.

halter73 replied:
These are all immutable, so structs are certainly an option. Given that all of these are used for requests that involve some sort of database interaction in the normal case, I don't think it will make a huge difference. It will matter more once we make these public (which I hope to do soon).

BrennanConroy commented on src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs:
```diff
@@ -42,44 +59,64 @@ public static IEndpointConventionBuilder MapIdentityApi<TUser>(this IEndpointRou
 
-        routeGroup.MapPost("/login", async Task<Results<UnauthorizedHttpResult, Ok<AccessTokenResponse>, SignInHttpResult>>
-            ([FromBody] LoginRequest login, [FromQuery] bool? cookieMode, [FromServices] IServiceProvider sp) =>
+        routeGroup.MapPost("/login", async Task<Results<Ok<AccessTokenResponse>, ProblemHttpResult, IResult>>
```
Should we put `EmptyHtppResult` here instead of `IResult`? `IResult` basically means anything can be returned from this endpoint?

halter73 replied:
It's not really empty. It's just not overwriting the AccessTokenResponse.

BrennanConroy commented on src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs:
```diff
@@ -95,20 +132,396 @@ await signInManager.ValidateSecurityStampAsync(refreshTicket.Principal) is not T
+            IdentityResult result;
+            try
+            {
+                code = Encoding.UTF8.GetString(WebEncoders.Base64UrlDecode(code));
```
Looks like something we could try optimizing in the future 😃

halter73 replied:
I'll accept PRs 😉

BrennanConroy commented on src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs:
```diff
@@ -95,20 +132,396 @@ await signInManager.ValidateSecurityStampAsync(refreshTicket.Principal) is not T
+                code = WebEncoders.Base64UrlEncode(Encoding.UTF8.GetBytes(code));
+
+                await emailSender.SendEmailAsync(resetRequest.Email, "Reset your password",
+                    $"Reset your password using the following code: {HtmlEncoder.Default.Encode(code)}");
```
I assume we'll want to localize these strings in the future?

halter73 replied:
Probably. It doesn't look like AddDefaultIdentity does in the [password reset razor page](https://github.com/dotnet/aspnetcore/blob/f9828517e57769959d43c89d06ad7f319880258c/src/Identity/UI/src/Areas/Identity/Pages/V5/Account/ForgotPassword.cshtml.cs#L84-L87), but it's worth considering. https://github.com/dotnet/aspnetcore/issues/49664

BrennanConroy commented on src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs:
```diff
@@ -95,20 +132,396 @@ await signInManager.ValidateSecurityStampAsync(refreshTicket.Principal) is not T
+                return TypedResults.NotFound();
+            }
+
+            var failedIdentityResults = new List<IdentityResult>();
```
nit: Don't allocate this until it's needed

BrennanConroy commented on src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs:
```diff
@@ -95,20 +132,396 @@ await signInManager.ValidateSecurityStampAsync(refreshTicket.Principal) is not T
+        if (string.IsNullOrEmpty(key))
+        {
+            await userManager.ResetAuthenticatorKeyAsync(user);
+            key = await userManager.GetAuthenticatorKeyAsync(user);
```
Can this still be `null` the second time?

halter73 replied:
The API signature says yes but ResetAuthenticatorKeyAsync succeeding says no. We should probably validate and actually check the IdentityResult returned by ResetAuthenticatorKeyAsync too. This does match the Identity UI behavior though.

BrennanConroy commented on src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs:
```diff
@@ -95,20 +132,396 @@ await signInManager.ValidateSecurityStampAsync(refreshTicket.Principal) is not T
+        return new()
+        {
+            Username = await userManager.GetUserNameAsync(user) ?? throw new NotSupportedException("Users must have a user name."),
+            Email = await userManager.GetEmailAsync(user) ?? throw new NotSupportedException("Users must have an email."),
```
Nit: run GetUserName and GetEmail before creating the claims dictionary.

BrennanConroy commented on src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs:
```diff
@@ -95,20 +132,396 @@ await signInManager.ValidateSecurityStampAsync(refreshTicket.Principal) is not T
         public void Finally(Action<EndpointBuilder> finallyConvention) => InnerAsConventionBuilder.Finally(finallyConvention);
     }
 
+    private sealed class NoopResult : IResult
```
Unused?

BrennanConroy commented on src/Identity/Extensions.Core/src/TokenProviderDescriptor.cs:
```diff
@@ -10,22 +11,40 @@ namespace Microsoft.AspNetCore.Identity;
 public class TokenProviderDescriptor
 {
+    // Temporary fix to test MapIdentityApi's support for multiple TUser and TContext.
+    // There's nothing as permanent as a temporary fix, but it seems better than now support.
```
> but it seems better than now support

"**new** support"?

halter73 replied:
```suggestion
    // There's nothing as permanent as a temporary fix, but it seems better than no support.
```

halter73 commented:
@BrennanConroy I responded to all of your feedback. Does this look good to merge?

@javiercn Let me know if you have any additional feedback. I'd like to merge this by tomorrow.

pinkfloydx33 commented on src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs:
```diff
@@ -95,15 +132,398 @@ await signInManager.ValidateSecurityStampAsync(refreshTicket.Principal) is not T
+                else if (!await userManager.VerifyTwoFactorTokenAsync(user, userManager.Options.Tokens.AuthenticatorTokenProvider, tfaRequest.TwoFactorCode))
+                {
+                    return CreateValidationProblem("InvalidTwoFactorCode",
+                        "The 2fa token provide by the request was invalid. A valid 2fa token is required to enable 2fa.");
```
```suggestion
                        "The 2fa token provided by the request was invalid. A valid 2fa token is required to enable 2fa.");
```

halter73 committed:
Update src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs

Co-authored-by: pinkfloydx33 <matt@mattseanhull.com>

augustevn commented:
I'm not sure where to ask this. I'm getting bad confirmation links from the Identity endpoints.
What am I missing to configure the base url? This is formed in the htmlMessage coming from the IEmailSender.

http://auth/confirmEmail?userId=6338db8a-02d0-4a48-9d32-fc18f20fb45a&code=Q2ZESjhBSEFKN1NWdWVGRnNBL1JDLzY4YitnaVcrdk91aGJyVFpqYTR5dy9PUFZhOW1ZbUpmalo3MWplVXZRYjNaK0ZqYnA5d0xveDk4a1hsbWFraHR1Ym9CYTZVZXdzZzVZQnFCbWpIYk9LaE5raWVGR2xYczBsSEdHYzFUVTZQaTRUVXp1ZExRRnNSSUhVYkV5RSsrUXgvWGR2T1hGclFCRTZoLzBaV0JxbDd1bFBZVDZNNFFiaGN1b0d1ZElEZFl3M0ZlUmZ6NklmL3FiaEo4VjJWZENhT29Fa1RrbEJrRlRqRS9UNk8vRTJMRjdNVjljU0grZk5tWU5Hc0p5UlIzd0xCQT09

Using Identity version 8.0.0-rc.1.23421.29
